### DISCUSSION TITLE
is there an event for "new session" on the engine?

### DISCUSSION THREAD

**Discussion by earonesty** [sqlalchemy/sqlalchemy#11215] [Usage Questions] [answered]
i'd like to listen to do_orm_execute for all sessions created from a speciic engine

i get that there's no event for do_orm_execute on the engine, since it's core, not orm

but when a sesssion is created, can i hook at that time?

**Comment by CaselIT** [comment]
Hi,

The easiest solution is to create a sessionmaker binded to that engine and listen to its event. https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.sessionmaker

**Comment by earonesty** [comment]
Yeah I don't own the session creation.  It's a monitor app for other users.  I'll just stick with requiring the user to inject .listen() calls on each session instead of the engine itself.

**Reply by CaselIT** [reply]
That's not very efficient. A better way is probably to do do something like the following (also considering that in the majority of cases a single engine would be in use)
```py
@listen(Session, 'do_orm_execute')
def fn(state):
  if state.session.bind is not my_engine:
    return
  ...
```

But if you can require users to add a listen to each session, then I would think you could require them to do the listen on the session maker.

**Reply by earonesty** [reply]
That's cool.  What is that decorator on a session class

**Reply by earonesty** [reply]
I thought the first argument to listen had to be an actual instance of something

**Reply by CaselIT** [reply, marked as answer]
No, classes like pool, engine, session (and sessionmaker I think) can also be used to indicate "all instances of this class"

**Reply by earonesty** [reply]
That's super cool.  Thanks.  Gets me everything I need.
